go/libkb/pvl_interp.go

```diff
@@ -0,0 +1,825 @@
+	if err != nil {
+		return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get length of script: %v", err)
+	}
+	if scriptlen < 1 {
+		return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Empty script")
+	}
+
+	for i := 0; i < scriptlen; i++ {
+		ins := script.AtIndex(i)
+		switch {
+
+		// These can always run, but must be cases so that the default case works.
+		case pvlJSONHasKey(ins, PvlAssertRegexMatch):
+		case pvlJSONHasKey(ins, PvlAssertFindBase64):
+		case pvlJSONHasKey(ins, PvlWhitespaceNormalize):
+		case pvlJSONHasKey(ins, PvlRegexCapture):
+
+		case pvlJSONHasKey(ins, PvlFetch):
+			// A script can contain only <=1 fetches.
+			// A DNS script cannot contain fetches.
+
+			fetchType, err := ins.AtKey(PvlFetch).GetString()
+			if err != nil {
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"Could not get fetch type %v", i)
+			}
+
+			if service == keybase1.ProofType_DNS {
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"DNS script cannot contain fetch instruction")
+			}
+			if modeknown {
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"Script cannot contain multiple fetch instructions")
+			}
+			switch PvlMode(fetchType) {
+			case PvlModeString:
+				modeknown = true
+				mode = PvlModeString
+			case PvlModeHTML:
+				modeknown = true
+				mode = PvlModeHTML
+			case PvlModeJSON:
+				modeknown = true
+				mode = PvlModeJSON
+			default:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"Unsupported fetch type: %v", fetchType)
+			}
+		case pvlJSONHasKey(ins, PvlSelectorJSON):
+			// Can only select after fetching.
+			switch {
+			case service == keybase1.ProofType_DNS:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"DNS script cannot use json selector")
+			case !modeknown:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"Script cannot select before fetch")
+			case mode != PvlModeJSON:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"Script contains json selector in non-html mode")
+			}
+		case pvlJSONHasKey(ins, PvlSelectorCSS):
+			// Can only select after fetching.
+			switch {
+			case service == keybase1.ProofType_DNS:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"DNS script cannot use css selector")
+			case !modeknown:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"Script cannot select before fetch")
+			case mode != PvlModeHTML:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"Script contains css selector in non-html mode")
+			}
+		case pvlJSONHasKey(ins, PvlTransformURL):
+			// Can only transform before fetching.
+			switch {
+			case service == keybase1.ProofType_DNS:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"DNS script cannot transform url")
+			case modeknown:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"Script cannot transform after fetch")
```

Comment:
Can scripts do a fetch, make some transformations/asserts, then do a different fetch, and make yet more transformations/asserts? Any reason to restrict it to one?

I can imagine keeping track of the active page in your head could get confusing, but I'm wondering if we've thought about it. The model we've built for ourselves here does seem like it could accommodate this sort of thing.